core/valor_core/utilities.py

```diff
@@ -0,0 +1,848 @@
+                    if col in df.columns
+                ]
+            ]
+            .notna()
+            .sum(axis=1)
+            == 1
+        )
+        & df.get("is_instance", null_placeholder_column).isin([True])
+        & df.get("embedding", null_placeholder_column).isnull()
+    ].index
+
+    # semantic segmentation tasks only support rasters
+    semantic_segmentation_rows = df[
+        df.get("label_key", null_placeholder_column).notnull()
+        & df.get("label_value", null_placeholder_column).notnull()
+        & df.get("bounding_box", null_placeholder_column).isnull()
+        & df.get("polygon", null_placeholder_column).isnull()
+        & df.get("raster", null_placeholder_column).notnull()
+        & df.get("embedding", null_placeholder_column).isnull()
+        & df.get("is_instance", null_placeholder_column).isin([None, False])
+    ].index
+
+    # empty annotations shouldn't contain anything
+    empty_rows = df[
+        df.get("label_key", null_placeholder_column).isnull()
+        & df.get("label_value", null_placeholder_column).isnull()
+        & df.get("bounding_box", null_placeholder_column).isnull()
+        & df.get("polygon", null_placeholder_column).isnull()
+        & df.get("raster", null_placeholder_column).isnull()
+        & df.get("embedding", null_placeholder_column).isnull()
+    ].index
+
+    if not classification_rows.empty:
+        df.loc[
+            classification_rows, "implied_task_type"
+        ] = enums.TaskType.CLASSIFICATION
+
+    if not object_detection_rows.empty:
+        df.loc[
+            object_detection_rows, "implied_task_type"
+        ] = enums.TaskType.OBJECT_DETECTION
+
+    if not semantic_segmentation_rows.empty:
+        df.loc[
+            semantic_segmentation_rows, "implied_task_type"
+        ] = enums.TaskType.SEMANTIC_SEGMENTATION
+
+    if not empty_rows.empty:
+        df.loc[empty_rows, "implied_task_type"] = enums.TaskType.EMPTY
+
+    if df["implied_task_type"].isnull().any():
+        raise ValueError(
+            "Input didn't match any known patterns. Classification tasks should only contain labels. Object detection tasks should contain labels and polygons, bounding boxes, or rasters with is_instance == True. Segmentation tasks should contain labels and rasters with is_instance != True. Text generation tasks should only contain text and optionally context."
+        )
+
+    return df
+
+
+def _convert_raster_to_box(raster: np.ndarray) -> geometry.Box:
+    """Convert a raster mask to a Box."""
+    rows = np.any(raster, axis=1)
+    cols = np.any(raster, axis=0)
+    if not np.any(rows) or not np.any(cols):
+        raise ValueError("Raster is empty, cannot create bounding box.")
+
+    ymin, ymax = np.where(rows)[0][[0, -1]]
+    xmin, xmax = np.where(cols)[0][[0, -1]]
+
+    return geometry.Box.from_extrema(xmin, xmax + 1, ymin, ymax + 1)
+
+
+def _convert_raster_to_polygon(raster: np.ndarray) -> geometry.Polygon:
```

Comment:
Is there a reference for the algorithm used here?

Reply:
there isn't. I need to test this further when adding better conversion tests see ("Issues to fix..")